src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php

```diff
@@ -312,33 +317,54 @@ private function getAccessorMethod($class, $property)
     private function getMutatorMethod($class, $property)
     {
         $ucProperty = ucfirst($property);
+        $ucSingulars = (array) Inflector::singularize($ucProperty);
 
         foreach (self::$mutatorPrefixes as $prefix) {
-            try {
-                $reflectionMethod = new \ReflectionMethod($class, $prefix.$ucProperty);
+            $names = array($ucProperty);
+            if (in_array($prefix, self::$arrayMutatorPrefixes)) {
+                $names = array_merge($names, $ucSingulars);
+            }
 
-                // Parameter can be optional to allow things like: method(array $foo = null)
-                if ($reflectionMethod->getNumberOfParameters() >= 1) {
-                    return array($reflectionMethod, $prefix);
+            foreach ($names as $name) {
+                try {
+                    $reflectionMethod = new \ReflectionMethod($class, $prefix.$name);
+
+                    // Parameter can be optional to allow things like: method(array $foo = null)
+                    if ($reflectionMethod->getNumberOfParameters() >= 1) {
+                        return array($reflectionMethod, $prefix);
+                    }
+                } catch (\ReflectionException $reflectionException) {
+                    // Try the next prefix if the method doesn't exist
                 }
-            } catch (\ReflectionException $reflectionException) {
-                // Try the next prefix if the method doesn't exist
             }
         }
     }
 
     /**
      * Extracts a property name from a method name.
      *
-     * @param string $methodName
+     * @param string                $methodName
+     * @param \ReflectionProperty[] $reflectionProperties
      *
      * @return string
      */
-    private function getPropertyName($methodName)
+    private function getPropertyName($methodName, array $reflectionProperties)
     {
         $pattern = implode('|', array_merge(self::$accessorPrefixes, self::$mutatorPrefixes));
 
         if (preg_match('/^('.$pattern.')(.+)$/i', $methodName, $matches)) {
+            if (!in_array($matches[1], self::$arrayMutatorPrefixes)) {
+                return $matches[2];
+            }
+
+            foreach ($reflectionProperties as $reflectionProperty) {
+                foreach ((array) Inflector::singularize($reflectionProperty->name) as $name) {
+                    if ($name === lcfirst($matches[2])) {
```

Comment:
Should we use `strtolower()` when comparing as method names are case-insensitive (like is the regular expression used at the beginning of this method)?